### PR TITLE
Added '--deps-only' to {build, doc} subcommands

### PR DESCRIPTION
**Pull request by KalitaAlexey**
Please don't merge it.
I haven't added tests.

It works, I just want to hear opinions.
Which subcommands can support "--deps-only"?

How can I prevent the bot from building?

Soon I will describe my changes.

#### What to do if both `--deps-only` and `--no-deps` are passed to `cargo doc`?

### CODE REVIEW

**Comment by KalitaAlexey** [comment]
@rust-highfive, stop build.

**Comment by KalitaAlexey** [comment]
@brson,
Hi. I want to write a test checking that there is a library in *deps* folder. The library has metadata-suffix.
How to deal with it?

**Comment by alexcrichton** [comment]
cc https://github.com/rust-lang/cargo/issues/2644 

I'm not personally sold on the motivation for such a feature here to add it to Cargo yet. Perhaps we could continue to gather use cases on the issue itself?

**Comment by KalitaAlexey** [comment]
@alexcrichton,
It is useful and easy to add. Why can't we?

**Comment by KalitaAlexey** [comment]
cc #2801

**Comment by alexcrichton** [comment]
The implementation is not as easy as you might think. For example the implementation here is incorrect. For example it looks like this will still compile the library of the current package if there's a binary.

I'd like to get more information about use cases before we add such a feature. "it is useful" to me isn't a concrete reason as to why this feature is motivated.

**Comment by KalitaAlexey** [comment]
@alexcrichton,

> For example it looks like this will still compile the library of the current package if there's a binary.

I can't understand that. Can you rephrase that please?

I will check whether it still compiles a library.

For me it is useful only for `cargo doc`, but some people also would like to have it for `cargo build`.

**Comment by alexcrichton** [comment]
@KalitaAlexey I'd encourage you to add some tests to this PR to see the behavior. You can try adding a project with both a library and a binary and then assert that `cargo build --no-deps` *doesn't* compile the library.

**Comment by KalitaAlexey** [comment]
@alexcrichton,
Did you mean `cargo build --deps-only`?

**Comment by alexcrichton** [comment]
Ah yes, I do

**Comment by KalitaAlexey** [comment]
Currently I little time.
When I have time I will improve this.

**Comment by alexcrichton** [comment]
Closing due to inactivity, but feel free to resubmit with comments addressed!

**Comment by atodorov** [comment]
@alexcrichton I'm a +1 for using this inside Docker but I'm not sure if we can do this. In my Dockerfile I have:
```
COPY src/ /bdcs-api-rs/
RUN cd /bdcs-api-rs/ && cargo build
```

What I want to do is:
```
COPY src/Cargo.toml /bdcs-api-rs/
RUN cd /bdcs-api-rs/ && cargo build --dependencies-only

COPY src/ /bdcs-api-rs/
RUN cd /bdcs-api-rs/ && cargo build
```

so that I rebuild the dependencies only if they changed and I can reuse the intermediate Docker container running the --dependencies-only step.

Is it possible to implement this option without having the source code available, only Cargo.toml and/or Cargo.lock ? If yes can you point to some places in the code I may be able to produce a PR ?

**Comment by alexcrichton** [comment]
@atodorov right now Cargo's not really built to cater to that use case unfortunately. Getting that working in a reliable fashion would be pretty difficult.

**Comment by atodorov** [comment]
FYI I've managed to work around this by using `cargo build -p` and a small Python script to parse `Cargo.toml`. Works well for me. If you need to see examples checkout my blog.

**Comment by nacardin** [comment]
I just published a cargo extension crate to make this easier. I've been using it in one of my projects. [https://crates.io/crates/cargo-build-deps](https://crates.io/crates/cargo-build-deps)
Feedback via issues/PRs would be appreciated. [https://github.com/nacardin/cargo-build-deps](https://github.com/nacardin/cargo-build-deps)
